File: LeetCode/202. Happy Number/202.cpp
class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> s;
        while (n != 1) {
            if (s.find(n) != s.end()) {
                return false;
            }
            s.insert(n);
            int sum = 0;
            while (n) {
                sum += (n % 10) * (n % 10);
                n /= 10;
            }
            n = sum;
        }
        return true;
        
    }
};
//other people's solution
/*
  class Solution {
  public:
      bool isHappy(int n) {
          std::unordered_map <int, int> mp;
          unsigned long sum = 0;
          while (true)
          {
              while (n > 0)
              {
                  sum += (n % 10)* (n % 10);
                  n /= 10;
              }
              if (sum == 1)
                  break;
              else if (mp[sum])
                  return false;
                n = sum;
                mp[sum]++;
                sum = 0;
          }
          return true;
      }
  };
*/

File: LeetCode/202. Happy Number/202.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> set = new HashSet<>();
        while (n != 1) {
            int sum = 0;
            while (n != 0) {
                int digit = n % 10;
                sum += digit * digit;
                n /= 10;
            }
            if (set.contains(sum)) {
                return false;
            }
            set.add(sum);
            n = sum;
        }
        return true;

    }

}

//other peoples solution
/*
 * import java.util.LinkedList;
 * class Solution {
 * public boolean isHappy(int n) {
 * int slow = n;
 * int fast = n;
 * do {
 * slow = square(slow);
 * fast = square(square(fast));
 * }
 * while (slow != fast);
 * return slow == 1;
 * }
 * public int square(int num) {
 * int ans = 0;
 * while(num > 0) {
 * int remainder = num % 10;
 * ans += remainder * remainder;
 * num /= 10;
 * }
 * return ans;
 * }
 * }
 */
